### DISCUSSION TITLE
connexion (version 2.14.2) is failing to extract parameter from multipart/form-data POST body with one string parameter and one file parameter

### DISCUSSION THREAD

**Discussion by pmajumdar1976** [spec-first/connexion#1679] [Q&A]
Here is a part of my openapi spec.

```
paths:
    /captures:
        description: A list of loaded captures
        post:
            operationId: captureLoad
            description: Load a new capture
            requestBody:
                description: The capture to be loaded
                required: true
                content:
                    multipart/form-data:
                        schema:
                            type: object
                            required:
                                - pcap_name
                                - pcap_file
                            properties:
                                pcap_name:
                                    type: string
                                    nullable: false
                                    example: Office365 Sample
                                pcap_file:
                                    type: string
                                    format: binary
                                    nullable: false
            responses:
                200:
                    description: The capture is loaded
                    content:
                        application/json:
                            schema:
                                $ref: '#/components/schemas/capture'
```
There are two required parameters in the requestBody. This format should be valid if the following example from https://swagger.io/docs/specification/describing-request-body/file-upload/ is valid.

```
      requestBody:
        content:
          multipart/form-data:
            schema:
              type: object
              properties:
                orderId:
                  type: integer
                userId:
                  type: integer
                fileName:
                  type: string
                  format: binary
```
The only difference in my spec is that I marked both of the properties as required.

The connexion module is failing to extract the **_pcap_name_** argument from the body. I am using openapi version 3.0.0. So the function **__get_body_argument_** from _connexion/operations/openapi.py_ is getting called.

Am I doing anything wrong here? If not then there is a bug in the **__get_body_argument_** function (at least in the version I am using).
Replacing the last line by **_return result_** (instead of **_return {})_** fixes the problem.

Please share your thoughts and let me know what the proper fix should be.

**_Is there some other version that I should upgrade to? I installed my version recently using pip install connexion._**

**Comment by pmajumdar1976** [comment]
I looked into the latest code in github. It may have a fix for the bug. The function **_prep_kwargs_** passes **_content_type=request.mimetype_** to **_get_arguments_** and eventually **__get_body_argument_** uses this **_content_type_**. This may work for my case. Is there any released version that I can install for trying this new code?

**Reply by RobbeSneyders** [reply]
Hi @pmajumdar1976,
Please check my explanation [here](url):

> You're right. With Swagger 2, this works since form data is defined as parameters in the specification. With OpenAPI 3, form data is defined in the body, so we pass in the whole body.
> 
> This will work:
> 
> ```
> def post(file, body):
>     # body = {"filetype": "..."}
> ```
> 
> But I agree that this is not very consistent. I think we should either
> 
> * Copy the Swagger behavior for OpenAPI and pass the top level fields in separately.
>   (This might impact the behavior of `strict_validation` as well.)
> * Don't pass in the files separately for OpenAPI either, and just include them in the body
> 
> I'll have to look how this was done on Connexion 2.X before deciding on how to move forward. CC: @Ruwann

**Reply by pmajumdar1976** [reply]
Hi @RobbeSneyders,

Thanks for looking into my question and responding. Unfortunately this doesn't solve my problem completely. Let me share a bit more information which will help you to understand my situation. I am using a popular tool called `openapitools/openapi-generator-cli` for generating server code from my openapi spec. It generates the signatures of the request handlers. I can manually change those files but I would like to avoid that.

The `openapitools/openapi-generator-cli` makes the required parameters the required arguments of the handler/controller (I think controller is the term used by `openapitools/openapi-generator-cli` tool) functions.

Interestingly `openapitools/openapi-generator-cli` uses this `connexion` module. That's why I started digging into it.

For now, I can just hack the `connexion` code in my local installation but I would like to hear your advice on it. Do you have any communication channel with the `openapitools/openapi-generator-cli` team? If you do then you may want to pick their brains too.
